### PR TITLE
feat: handle theme data from content and local storage

### PR DESCRIPTION
**Pull request by briantstephan**
This gets the theme data from content/local storage and sends it along to the theme sidebar so that the sidebar is up to date. It also handles saving by sending saved/published changes up to the parent.

Can't test until the data is sent down from the parent.

### CODE REVIEW

File: src/components/Editor.tsx [hunk 1]
```diff
@@ -69,7 +84,7 @@ export const Editor = ({
     }
 
     return getLocalStorageKey(
-      !!templateMetadata.isThemeMode,
+      templateMetadata.isThemeMode,
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:87]: Why is theme mode relevant to the non-theme key? We have a separate buildThemeKey func now

File: src/components/Editor.tsx [hunk 2]
```diff
@@ -120,21 +147,38 @@ export const Editor = ({
     }
   }, [templateMetadata?.isDevMode]);
 
+  const clearLocalStorage = () => {
+    if (!templateMetadata?.isThemeMode) {
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:151]: Prefer the positive, so flip this around to `if true clearTheme()`

File: src/components/Editor.tsx [hunk 3]
```diff
@@ -120,21 +147,38 @@ export const Editor = ({
...
+      clearThemeLocalStorage();
+    }
+  };
+
   /**
    * Clears the user's localStorage and resets the current Puck history
    */
-  const clearLocalStorage = () => {
-    devLogger.logFunc("clearLocalStorage");
+  const clearPuckLocalStorage = () => {
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:161]: clearVisualConfigLocalStorage?

File: src/components/Editor.tsx [hunk 4]
```diff
@@ -120,21 +147,38 @@ export const Editor = ({
...
+    devLogger.logFunc("clearPuckLocalStorage");
     window.localStorage.removeItem(buildLocalStorageKey());
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:163]: Should we update the storage key func name too?

File: src/components/Editor.tsx [hunk 5]
```diff
@@ -144,7 +188,13 @@ export const Editor = ({
       return;
     }
     loadPuckInitialHistory(); // do something after state has updated
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:190]: Can you remove this comment? It's not helpful

File: src/components/Editor.tsx [hunk 6]
```diff
@@ -349,16 +526,31 @@ export const Editor = ({
...
+    "deleteThemeSaveState",
+    TARGET_ORIGINS
+  );
+
   const { sendToParent: saveVisualConfigData } = useSendMessageToParent(
     "saveVisualConfigData",
     TARGET_ORIGINS
   );
 
+  const { sendToParent: saveThemeConfigData } = useSendMessageToParent(
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:549]: Group all of the theme stuff together

File: src/components/Editor.tsx [hunk 7]
```diff
@@ -383,7 +575,11 @@ export const Editor = ({
...
     !visualConfigurationDataFetched ||
-    !puckInitialHistoryFetched;
+    !puckInitialHistoryFetched ||
+    (templateMetadata.isThemeMode &&
+      (!themeDataFetched ||
+        !themeSaveStateFetched ||
+        !themeInitialHistoryFetched));
 
   const progress: number =
     (100 * // @ts-expect-error adding bools is fine
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:585]: Progress % needs to correspond with the things guarding isLoading

File: src/components/Editor.tsx [hunk 8]
```diff
@@ -78,13 +78,12 @@ export const Editor = ({
     devLogger.logData("DOCUMENT", document);
   }
 
-  const buildLocalStorageKey = useCallback(() => {
-    if (!templateMetadata) {
+  const buildVisualConfigLocalStorageKey = useCallback(() => {
+    if (!templateMetadata || templateMetadata.isThemeMode) {
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:82]: Don't we still need the visualConfiguration data from localStorage even if you're in theme mode?

**Comment by jwartofsky-yext** [comment]
Data should be sent down from the parent now

File: src/components/Editor.tsx (hunk 8 above)
Comment (briantstephan) [reply src/components/Editor.tsx]: Don't think so, I just prevented those fetches from occurring when in theme mode

File: src/components/Editor.tsx [hunk 9]
```diff
@@ -210,7 +210,8 @@ export const Editor = ({
     if (
       !visualConfigurationDataFetched ||
       !saveStateFetched ||
-      !templateMetadata
+      !templateMetadata ||
+      templateMetadata.isThemeMode
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:214]: We need this otherwise nothing will render in the Puck view

File: src/components/Editor.tsx [hunk 10]
```diff
@@ -463,6 +464,9 @@ export const Editor = ({
   }, [templateMetadata]);
 
   useReceiveMessage("getSaveState", TARGET_ORIGINS, (send, payload) => {
+    if (templateMetadata?.isThemeMode) {
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:467]: We need this too

File: src/components/Editor.tsx [hunk 11]
```diff
@@ -483,6 +487,9 @@ export const Editor = ({
     "getVisualConfigurationData",
     TARGET_ORIGINS,
     (send, payload) => {
+      if (templateMetadata?.isThemeMode) {
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:490]: And this

File: src/components/Editor.tsx [hunk 12]
```diff
@@ -574,9 +581,10 @@ export const Editor = ({
     !puckConfig ||
     !templateMetadata ||
     !document ||
-    !saveStateFetched ||
-    !visualConfigurationDataFetched ||
-    !puckInitialHistoryFetched ||
+    (!templateMetadata.isThemeMode &&
```
Comment (mkilpatrick) [inline src/components/Editor.tsx:584]: Revert this

File: src/internal/puck/components/Header.tsx [hunk 13]
```diff
@@ -78,7 +78,7 @@ export const customHeader = (
           </Button>
         )}
         <ClearLocalChangesButton
-          disabled={histories.length === 1}
+          disabled={isThemeMode ? false : histories.length === 1}
```
Comment (mkilpatrick) [inline src/internal/puck/components/Header.tsx:81]: isThemeMode ? theme.histories.length === 1 : histories.length === 1, same below

File: src/components/Editor.tsx (hunk 8 above)
Comment (mkilpatrick) [reply src/components/Editor.tsx:82]: None of the VConfig stuff should be affected by theme. We need it always.

**Review by mkilpatrick — APPROVED** [review]

**Pull request merged**